packages/editor-ui/src/components/RunData.vue

```diff
@@ -578,14 +603,50 @@ export default mixins(
 					return;
 				}
 
-				this.editMode.enabled = false;
-				this.$store.commit('pinData', { node: this.node, data: this.editMode.value });
+				try {
+					if (this.$store.getters['pinDataSize'] + stringSizeInBytes(JSON.stringify(data)) > MAX_WORKFLOW_PINNED_DATA_SIZE) {
+						throw new Error(this.$locale.baseText('ndv.pinData.error.tooLarge.description'));
+					}
+				} catch (error) {
+					this.$showError(error, this.$locale.baseText('ndv.pinData.error.tooLarge.title'));
+					return;
+				}
+
+				if (this.verifyPinnedDataSize(data)) {
+					this.editMode.enabled = false;
+					this.$store.commit('pinData', { node: this.node, data });
+				}
 			},
-			onTogglePinData() {
+			async onTogglePinData() {
 				if (this.hasPinData) {
 					this.$store.commit('unpinData', { node: this.node });
-				} else {
+				} else if (this.verifyPinnedDataSize(this.jsonData)) {
 					this.$store.commit('pinData', { node: this.node, data: this.jsonData });
+
+					if (this.maxRunIndex > 0) {
+						this.$showToast({
+							title: this.$locale.baseText('ndv.pinData.pin.multipleRuns.title', {
+								interpolate: {
+									index: `${this.runIndex}`,
+								},
+							}),
+							message: this.$locale.baseText('ndv.pinData.pin.multipleRuns.description'),
+							type: 'success',
+							duration: 2000,
+						});
+					}
+				}
+			},
+			verifyPinnedDataSize(data: IDataObject[]): boolean {
+				try {
+					if (this.$store.getters['pinDataSize'] + stringSizeInBytes(JSON.stringify(data)) > MAX_WORKFLOW_PINNED_DATA_SIZE) {
+						throw new Error(this.$locale.baseText('ndv.pinData.error.tooLarge.description'));
+					}
+
+					return true;
+				} catch (error) {
+					this.$showError(error, this.$locale.baseText('ndv.pinData.error.tooLarge.title'));
+					return false;
```

Comment:
1. Looks like whenever you call `stringSizeInBytes` you need to stringify first. Maybe we should stringify inside the function.
2. Instead of a try-catch, I think this would be more readable with an if check.
3. Since it returns a boolean, naming: `isValidPinDataSize`?

Reply:
- Changed to if check instead of try-catch. 
- Renamed to `isValidPinDataSize`. 
- Now accepting both strings and objects, and stringifying values if they're not of string type